DRAFT Feature/advanced find

Atlas139mkm opened the pull request:
<!--

### Requirements

* Filling out the template is required. Any pull request that does not include
enough information to be reviewed in a timely manner may be closed at the
maintainers' discretion.
* All new code requires unit tests to ensure they work as expected and will
continue to work as new code is added in the future (regression testing).
* Make sure your branch name is prefixed by `feature`, `bugfix` or `release`
* Have you read Constellation's Code of Conduct? By filing an issue, you are
expected to comply with it, including treating everyone with respect:
https://github.com/constellation-app/constellation/blob/master/CODE_OF_CONDUCT.md

-->

### Prerequisites

- [ ] Reviewed the [checklist](CHECKLIST.md)

- [ ] Reviewed feedback from the "Sonar Cloud" bot. Note that you have to wait
    for the "CI / Unit Tests") to complete first.

### Description of the Change

This PR contains the UI elements for the Advanced Find View Tab. It contains several classes which make up separate parts of the bigger picture. There is no searching logic developed yet however that will be the next step after reviewing feedback. The changes will best be shown with screenshots provided below.

Default Smallest Size.
![image](https://user-images.githubusercontent.com/88649439/143326102-c32a4503-f61c-4c62-9fb4-22f6e214dfdd.png)

Showing off in order panel: String, Float, Icon, Color, Boolean
![image](https://user-images.githubusercontent.com/88649439/143326224-6a2ea7c8-5481-4731-b4ae-a1be213cea6a.png)

DateTime and DateTimeSelector
![image](https://user-images.githubusercontent.com/88649439/143326266-c8c2b7b9-904d-42f1-afe1-f073b39a2148.png)

String and UseListWindow
![image](https://user-images.githubusercontent.com/88649439/143326308-99c56afa-0c51-4ceb-800d-3b1cdc9d164a.png)

Icon and IconSelector
![image](https://user-images.githubusercontent.com/88649439/143326362-6ef8aca8-51d9-4504-b8b3-0316a1475be9.png)


<!--

We must be able to understand the design of your change from this description.
If we can't get a good idea of what the code will be doing from the description
here, the pull request may be closed at the maintainers' discretion. Keep in
mind that the maintainer reviewing this PR may not be familiar with or have
worked with the code here recently, so please walk us through the concepts.

-->

### Alternate Designs

This topic is still up for debate however the current design should maintain a similar feel to the existing advanced search while being in a more compact version. Implementations of alternate design will get reviewed once user testing has commenced.

<!--

Explain what other alternates were considered and why the proposed version was
selected.

-->

### Why Should This Be In Core?

This is a request to move this into the find view rewrite.

<!--

Explain why this functionality should be in Constellation Core as opposed to a
different module suite. Note that this question is more applicable when adding
new functionality. If this change is a minor update to an existing file then it
is understood that this change has to be to this module suite and a response
and therefore a response to this question is not required.

-->


### Possible Drawbacks

I wasn't sure how to do it when writing it but maybe someone can shed some light. My popup windows for String, DateTime and Icon are currently set to always on top. I'm aware this is not a good thing however I wasn't sure how to set it so it forced the user to interact with them before moving on. So if anyone knows what preference I need to turn on to do that. It would be greatly appreciated.

<!-- What are the possible side-effects or negative impacts of the code change? -->

### Verification Process

I have done a lot of my own user testing to ensure that nothing was breaking and code was doing as it was intended. There is a chance I have missed as its also not a fully furnished product yet. There is no unit tests yet either.

<!--

What process did you follow to verify that your change has the desired effects?

- How did you verify that all new functionality works as expected?
- How did you verify that all changed functionality works as expected?
- How did you verify that the change has not introduced any regressions?

Describe the actions you performed (e.g. buttons you clicked, text you typed,
commands you ran, etc.), and describe the results you observed.

-->

### Applicable Issues
#1489 


<!-- Link any applicable issues here -->

antares1470 commented:
@Atlas139mkm are you able to merge master into your other branch? Just noting it appears you have done so in this branch but not the other one and so the diff appears larger than I think it actually is

Atlas139mkm commented:
Sorry about that @antares1470, It should look a bit better now

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/DateTimeSelector.java:
```diff
@@ -0,0 +1,259 @@
+        // the timeZoneComboBox
+        timeZoneComboBox = new ComboBox<>();
+        final List<ZoneId> timeZoneList = timeZones.sorted(zoneIdComparator);
+        for (ZoneId id : timeZoneList) {
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/IconNode.java:
```diff
@@ -0,0 +1,57 @@
+    private ConstellationIcon icon;
+    private String iconName;
+
+    public IconNode(ConstellationIcon icon) {
```
parameter final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/IconSelector.java:
```diff
@@ -0,0 +1,197 @@
+         * is Flag.Australia. Flag will be saved as a category.
+         */
+        final Set<String> categories = new LinkedHashSet<>();
+        for (ConstellationIcon icon : IconManager.getIcons()) {
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/IconSelector.java:
```diff
@@ -0,0 +1,197 @@
+         * For each of the categories categories create a TreeItem and add it to
+         * rootNode
+         */
+        for (String categoryName : categories) {
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/IconSelector.java:
```diff
@@ -0,0 +1,197 @@
+         * categoryName add it to the list of iconNames.
+         */
+        final List<String> iconNames = new ArrayList<>();
+        for (ConstellationIcon icon : IconManager.getIcons()) {
```
similar in this function

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java:
```diff
@@ -0,0 +1,203 @@
+        setContent();
+        setAlwaysOnTop(true);
+
+        splitLinesChoiceBox.getSelectionModel().selectedItemProperty().addListener((ObservableValue<? extends String> observableValue, final String oldElement, final String newElement) -> {
```
lambda parameters final (or types removed)

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java:
```diff
@@ -0,0 +1,203 @@
+     */
+    public void updateText(String text) {
+        StringBuilder sb = new StringBuilder();
+        String[] splitText = text.split(",");
```
use `SeparatorConstants.COMMA`

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java:
```diff
@@ -0,0 +1,203 @@
+        String[] splitText = text.split(",");
+        for (int i = 0; i < splitText.length; i++) {
+            sb.append(splitText[i]);
+            sb.append(i == splitText.length - 1 ? "" : "\n");
```
`SeparatorConstants.NEWLINE`

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java:
```diff
@@ -0,0 +1,203 @@
+        String splitAt;
+        switch (selectedSplitChoice) {
+            case "Comma":
+                splitAt = ",";
```
Use `SeparatorConstants` for these `splitAt` sets (there's one for semicolon and tab as well)

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java:
```diff
@@ -0,0 +1,203 @@
+     * selected. The text are will now contain This\nIs\n\ntest
+     */
+    private void splitLinesAction() {
+        String selectedSplitChoice = splitLinesChoiceBox.getSelectionModel().getSelectedItem();
+        String splitAt;
```
both can be final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java:
```diff
@@ -0,0 +1,203 @@
+                splitAt = ",";
+                break;
+        }
+        String[] splitTextField = textArea.getText().split(splitAt);
+        textArea.clear();
+        StringBuilder sb = new StringBuilder();
```
finals

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/StringCriteriaPanel.java:
```diff
@@ -0,0 +1,153 @@
+     */
+    public void setSearchFieldText(final String text) {
+        final StringBuilder sb = new StringBuilder();
+        final String[] splitText = text.split("\n");
```
`SeparatorConstants.NEWLINE`

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java:
```diff
@@ -0,0 +1,203 @@
+        StringBuilder sb = new StringBuilder();
+        for (int i = 0; i < splitTextField.length; i++) {
+            sb.append(splitTextField[i]);
+            sb.append(i == splitTextField.length - 1 ? "" : "\n");
```
`SeparatorConstants.NEWLINE`

antares1470 reviewed (changes requested):
Gave the advanced find a test. I couldn't seem to get the search to work. I tried a basic search like 'Identifier is blah' but couldn't get the results to be selected. Tried it in the basic find and was able to confirm it selected there

Atlas139mkm commented:
@antares1470 that has not been implemented yet. This initial PR was just for the UI elements, the searching and selecting is what will come next.

antares1470 commented:
> @antares1470 that has not been implemented yet. This initial PR was just for the UI elements, the searching and selecting is what will come next.

Ah right. Missed that mention in the description. As you were

Atlas139mkm commented:
**Current situation:**

**bugs:**
- no known bugs

**Issues:**
- pop up editor windows (icon selector, date selector, useList string editor) do not force the user to interact with them before moving on. (I'm not sure how make them do that)

**new additions in latest commit:**
- Advanced find now works! added all logic to ensure search works based off users input
- Attributes lists update when an attribute is added
- Advanced find tab Attributes are now listed in Alphabetical order
- Fixed a bug with the string panel UseList popup
- Fixed a UI bug with the stringCriteraPanel
- Icon selector now sorts categories alphabetically
- Icon selector now opens to the currently selected Icon
- ColourCriteriaPanel now contains a constellation named colour picker (exactly like the on in the attribute editor)

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/StringCriteriaPanel.java:
```diff
@@ -0,0 +1,181 @@
+    public FindCriteriaValues getCriteriaValues() {
+        // if use list is selected, retrieve all values within the list
+        if (useListCheckBox.isSelected()) {
+            String[] splitStrings = searchField.getText().split(SeparatorConstants.COMMA);
+            List<String> stringList = new ArrayList<>();
```
both can be final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/criteriavalues/BooleanCriteriaValues.java:
```diff
@@ -0,0 +1,35 @@
+
+    private final boolean boolValue;
+
+    public BooleanCriteriaValues(String attributeType, String attribute, String filter, boolean boolValue) {
```
parameters final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/criteriavalues/ColourCriteriaValues.java:
```diff
@@ -0,0 +1,37 @@
+
+    private final ConstellationColor color;
+
+    public ColourCriteriaValues(String attributeType, String attribute, String filter, ConstellationColor color) {
```
similar

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/criteriavalues/DateTimeCriteriaValues.java:
```diff
@@ -0,0 +1,47 @@
+    private final String dateTimeStringPrimary;
+    private final String dateTimeStringSecondary;
+
+    public DateTimeCriteriaValues(String attributeType, String attribute, String filter, String dateTimeStringPrimary) {
```
similar for this and remaining criteria values

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/AdvancedSearchParameters.java:
```diff
@@ -0,0 +1,224 @@
+     * @param currentSelection
+     * @param searchAllGraphs
+     */
+    public AdvancedSearchParameters(final List<FindCriteriaValues> criteriaValuesList, final GraphElementType graphElementType, String allOrAny, String currentSelection, final boolean searchAllGraphs) {
```
parameters final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/AdvancedSearchParameters.java:
```diff
@@ -0,0 +1,224 @@
+     * @return
+     */
+    @Override
+    public boolean equals(final Object object) {
```
a few finals can be added to variables in the function code

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/AdvancedSearchParameters.java:
```diff
@@ -0,0 +1,224 @@
+        int i = 0;
+        if (criteriaValuesList.size() == parameters.getCriteriaValuesList().size()) {
+            for (FindCriteriaValues values : criteriaValuesList) {
+                if (!values.getAttribute().equals(parameters.getCriteriaValuesList().get(i).getAttribute())) {
+                    return false;
+                }
+                if (!values.getAttributeType().equals(parameters.getCriteriaValuesList().get(i).getAttributeType())) {
+                    return false;
+                }
+                if (!values.getFilter().equals(parameters.getCriteriaValuesList().get(i).getFilter())) {
```
these ifs can be merged with ||

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/AdvancedSearchParameters.java:
```diff
@@ -0,0 +1,224 @@
+                            StringCriteriaValues stringParameterValues = (StringCriteriaValues) values;
+                            StringCriteriaValues stringActualValues = (StringCriteriaValues) parameters.getCriteriaValuesList().get(i);
+
+                            if (!stringParameterValues.isIgnoreCase() == stringActualValues.isIgnoreCase()) {
+                                return false;
+                            }
+                            if (!stringParameterValues.isUseList() == stringActualValues.isUseList()) {
+                                return false;
+                            }
+                            if (!stringActualValues.isUseList() && !stringParameterValues.getText().equals(stringActualValues.getText())) {
+                                return false;
+                            }
+                            if (stringActualValues.isUseList() && !stringParameterValues.getTextList().equals(stringActualValues.getTextList())) {
```
similar

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/AdvancedSearchParameters.java:
```diff
@@ -0,0 +1,224 @@
+                        case FloatAttributeDescription.ATTRIBUTE_NAME:
+                            FloatCriteriaValues floatParameterValues = (FloatCriteriaValues) values;
+                            FloatCriteriaValues floatActualValues = (FloatCriteriaValues) parameters.getCriteriaValuesList().get(i);
+                            if (floatParameterValues.getFloatValuePrimary() != floatActualValues.getFloatValuePrimary()) {
+                                return false;
+                            }
+                            if (floatParameterValues.getFloatValueSecondary() != floatActualValues.getFloatValueSecondary()) {
```
similar

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/AdvancedSearchParameters.java:
```diff
@@ -0,0 +1,224 @@
+                        case ZonedDateTimeAttributeDescription.ATTRIBUTE_NAME:
+                            DateTimeCriteriaValues dateTimeParameterValues = (DateTimeCriteriaValues) values;
+                            DateTimeCriteriaValues dateTimeActualValues = (DateTimeCriteriaValues) parameters.getCriteriaValuesList().get(i);
+                            if (!dateTimeParameterValues.getDateTimeStringPrimaryValue().equals(dateTimeActualValues.getDateTimeStringPrimaryValue())) {
+                                return false;
+                            }
+                            if (!dateTimeParameterValues.getDateTimeStringSecondaryValue().equals(dateTimeActualValues.getDateTimeStringSecondaryValue())) {
```
similar

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/IconSelector.java:
```diff
@@ -0,0 +1,214 @@
+        final Set<String> sortedCategories = new LinkedHashSet<>();
+        sortedCategories.addAll(categories);
+
+        String[] iconName = icon.getExtendedName().split("\\.");
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/IconSelector.java:
```diff
@@ -0,0 +1,214 @@
+        }
+        // sort the icons in alpabetical order
+        sort(iconNames);
+        String[] iconName = icon.getExtendedName().split("\\.");
+
+        // for each of those iconNames find the icon and create an IconNode
+        for (String name : iconNames) {
+            IconNode node = new IconNode(IconManager.getIcon(name));
```
finals

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java:
```diff
@@ -0,0 +1,206 @@
+        bp.setBottom(buttonsBp);
+
+        bp.setCenter(vbox);
+        Scene scene = new Scene(bp);
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/advanced/utilities/UseListInputWindow.java:
```diff
@@ -0,0 +1,206 @@
+     * @param text
+     */
+    public void updateText(String text) {
+        StringBuilder sb = new StringBuilder();
+        String[] splitText = text.split(SeparatorConstants.COMMA);
```
finals

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/advanced/AdvancedSearchPlugin.java:
```diff
@@ -0,0 +1,717 @@
+
+    private static final int STARTING_INDEX = -1;
+
+    public AdvancedSearchPlugin(final AdvancedSearchParameters parameters, boolean selectAll, boolean selectNext) {
```
parameters final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/advanced/AdvancedSearchPlugin.java:
```diff
@@ -0,0 +1,717 @@
+            final FindResultsList oldList = new FindResultsList(STARTING_INDEX, foundResult.getAdvancedSearchParameters(), foundResult.getGraphId());
+            final AdvancedSearchParameters oldparameters = oldList.getAdvancedSearchParameters();
+            final AdvancedSearchParameters newParamters = new AdvancedSearchParameters(this.parameters);
+            int newIndex = getIndex(newParamters, oldparameters, foundResult.getCurrentIndex());
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/advanced/AdvancedSearchPlugin.java:
```diff
@@ -0,0 +1,717 @@
+            clearSelection(graph);
+        }
+        findInCurrentSelectionList = new FindResultsList(graph.getId());
+        FindResultsList findAllMatchingResultsList = new FindResultsList(graph.getId());
```
final. There's a few finals missing in this function but I won't be through all of them

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/advanced/AdvancedSearchPlugin.java:
```diff
@@ -0,0 +1,717 @@
+     * @return
+     */
+    private boolean searchAsString(final FindCriteriaValues values, final int attributeInt, final int currElement, final GraphWriteMethods graph) {
+        StringCriteriaValues stringValues = (StringCriteriaValues) values;
```
a bunch of missing finals in this function as well

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/advanced/AdvancedSearchPlugin.java:
```diff
@@ -0,0 +1,717 @@
+     * @return
+     */
+    private boolean searchAsFloat(final FindCriteriaValues values, final int attributeInt, final int currElement, final GraphWriteMethods graph) {
+        FloatCriteriaValues floatValues = (FloatCriteriaValues) values;
```
final

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/advanced/AdvancedSearchPlugin.java:
```diff
@@ -0,0 +1,717 @@
+         * date time values
+         */
+        switch (dateTimeValues.getFilter()) {
+            case "Occoured On":
```
Occur rather than Occour (if that is what it is elsewhere, that should probably be fixed there too)

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/plugins/advanced/AdvancedSearchPlugin.java:
```diff
@@ -0,0 +1,717 @@
+     * @return
+     */
+    private boolean searchAsDateTime(final FindCriteriaValues values, final int attributeInt, final int currElement, final GraphWriteMethods graph) {
+        DateTimeCriteriaValues dateTimeValues = (DateTimeCriteriaValues) values;
```
finals in this function as well

antares1470 commented on CoreFindView/src/au/gov/asd/tac/constellation/views/find2/components/AdvancedFindTab.java:
```diff
@@ -97,25 +96,18 @@
-                updateSelectionFactors();
-            }
-        });
+        currentSelectionChoiceBox.getSelectionModel().selectedItemProperty().addListener(action -> updateSelectionFactors());
```
I think the lambda needs to have 3 parameters rather than just 1 to match what the anonymous function had. Else I think you're adding an invalidation listener rather than a change listener

Unless of course that was intentional in which case feel free to ignore this suggestion

Atlas139mkm replied:
Hey @antares1470 are you able to clarify what you mean here? The intended action of the line is to just call updateSelectionFactors() when the selection changes.

antares1470 replied:
@Atlas139mkm so you're previous inner class function had 3 parameters: `observableValue`, `oldElement`, and `newElement`. The names aren't necessarily important here but point is there are 3. When you convert to a lambda function, the parameters of the lambda function should match those of that function as that is how it determines what is being added (in this case a `ChangeListener`). 

With your lambda function though, it only has one parameter. Consequently, rather than mapping to a `ChangeListener`, it is mapping to an `InvalidationListener` (whose function only has one parameter) which is the other kind of listener you can add here. Both kinds of listeners respond to different things and so consequently your code behaviour has changed, and by the sounds of it away from the intended behaviour.

Does that makes sense?
